filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobFileSystemManager.scala

```diff
@@ -169,54 +179,81 @@ object BlobSasTokenGenerator {
     * @return A WSMBlobTokenGenerator, able to produce a valid SAS token for accessing the provided blob
     * container and endpoint that is managed by WSM
     */
-  def createBlobTokenGenerator(container: BlobContainerName,
-                               endpoint: EndpointURL,
-                               workspaceId: WorkspaceId,
-                               containerResourceId: ContainerResourceId,
-                               workspaceManagerClient: WorkspaceManagerApiClientProvider,
-                               overrideWsmAuthToken: Option[String]): BlobSasTokenGenerator = {
-    WSMBlobSasTokenGenerator(container, endpoint, workspaceId, containerResourceId, workspaceManagerClient, overrideWsmAuthToken)
+  def createBlobTokenGenerator(workspaceManagerClient: WorkspaceManagerApiClientProvider,
+                               overrideWsmAuthToken: Option[String],
+                               buffer: Duration): BlobSasTokenGenerator = {
+    new WSMBlobSasTokenGenerator(workspaceManagerClient, overrideWsmAuthToken, buffer)
   }
 
 }
 
-case class WSMBlobSasTokenGenerator(container: BlobContainerName,
-                                    endpoint: EndpointURL,
-                                    workspaceId: WorkspaceId,
-                                    containerResourceId: ContainerResourceId,
-                                    wsmClientProvider: WorkspaceManagerApiClientProvider,
-                                    overrideWsmAuthToken: Option[String]) extends BlobSasTokenGenerator {
+class WSMBlobSasTokenGenerator(wsmClientProvider: WorkspaceManagerApiClientProvider,
+                                    overrideWsmAuthToken: Option[String],
+                                    buffer: Duration) extends BlobSasTokenGenerator {
 
+  var cachedSasTokens : TrieMap[(EndpointURL, BlobContainerName), AzureSasCredential] = new TrieMap();
+  def getAvailableCachedSasToken(endpoint: EndpointURL, container: BlobContainerName) = cachedSasTokens.get((endpoint, container))
+  def putAvailableCachedSasToken(endpoint: EndpointURL, container: BlobContainerName, sas: AzureSasCredential): Unit = {
+    cachedSasTokens.addOne(((endpoint, container), sas))
+    ()
+  }
   /**
-    * Generate a BlobSasToken by using the available authorization information
+    * Fetch a BlobSasToken from a cache or fall back to using the available authorization information
     * If an overrideWsmAuthToken is provided, use this in the wsmClient request
     * Else try to use the environment azure identity to request the SAS token
     *
     * @return an AzureSasCredential for accessing a blob container
     */
-  def generateBlobSasToken: Try[AzureSasCredential] = {
+  def findBlobSasToken(endpoint: EndpointURL, container: BlobContainerName, buffer: Duration): Try[AzureSasCredential] = {
```

Comment:
We're still in the `synchronized` block here, right? Might be worth adding a comment that we assume thread safety is being taken care of elsewhere.